### PR TITLE
Added ExtendTci check on Derived Context

### PR DESCRIPTION
**Pull request by hpya93**
@jhand2 
Hi, The commit reg removal of GetTaggedTci, TagTCI (PR257) removed the testcase for ExtendTci test for derived context. 
Because it was using TaggedTci and GetTagged Tci commands to crosscheck cumulative measurement.
The test was earlier reviewed and approved in PR 253.
Have rewritten the same test to verify ExtendTci on derived context but this time it uses CertifyKey instead of TagTci/GetTaggedTci
Can you please have a look at this PR ?

### CODE REVIEW

File: verification/certifyKey.go [hunk 1]
```diff
@@ -630,3 +668,58 @@ func getKeyUsageNames(keyUsage x509.KeyUsage) []string {
...
+	}
+
+	if !(pubKeyInResponse.Equal(pubKeyInCert)) {
+		t.Errorf("[ERROR]: Public key returned in response must match the Public Key Info in the certificate.")
+	}
+}
+
+// Checks whether the context handle is unchanged after certifyKey command when default context handle is used.
+func checkCertifyKeyRespHandle(res CertifiedKey, t *testing.T, handle *ContextHandle) {
+	if *handle != DefaultContextHandle {
```
Comment (hpya93) [inline verification/certifyKey.go:717]: A simple check to validate response handle

File: verification/certifyKey.go [hunk 2]
```diff
@@ -630,3 +668,58 @@ func getKeyUsageNames(keyUsage x509.KeyUsage) []string {
 
 	return keyUsageNames
 }
+
+// Checks CertifyKey command response against public key extracted from certificate returned in response
+func checkCertifyKeyResponse(t *testing.T, x509Cert *x509.Certificate, response CertifiedKey, hashAlg asn1.ObjectIdentifier) {
+	var err error
```
Comment (hpya93) [inline verification/certifyKey.go:674]: Checks whether public key returned in response of CertifyKey matches the key in DPE cert

File: verification/certifyKey.go [hunk 3]
```diff
@@ -249,42 +262,55 @@ func checkCertifyKeyMultiTcbInfoExtensionStructure(t *testing.T, c *x509.Certifi
 // Checks the FWID block's Digest.
 // FWID at index 0 has the TCI_CURRENT as digest
 // FWID at index 1 has the TCI_CUMULATIVE as digest
-// The length of FWID array in each DICE TCB information block is 2.
-func checkCurrentDiceTcbMeasurements(t *testing.T, multiTcbInfo []DiceTcbInfo, expectedCurrentValue []byte) {
-	currentTci := multiTcbInfo[0].Fwids[0].Digest
-	cumulativeTci := multiTcbInfo[0].Fwids[1].Digest
+// The length of FWID array in each DICE TCB information block is alyays 2 comprising the current and cumulative.
+func checkCurrentDiceTcbMeasurements(t *testing.T, handle *ContextHandle, multiTcbInfo []DiceTcbInfo, expectedCurrentValue []byte, isTciExtended bool) {
+	t.Helper()
```
Comment (hpya93) [inline verification/certifyKey.go:267]: Modified to validate when TCI value is extended.

File: verification/certifyKey.go [hunk 4]
```diff
@@ -120,18 +123,23 @@ const (
...
+	Label []byte
+	Flags CertifyKeyFlags
+}
+
 func TestCertifyKey(d TestDPEInstance, c DPEClient, t *testing.T) {
 	testCertifyKey(d, c, t, false)
 }
 
-func TestCertifyKey_SimulationMode(d TestDPEInstance, c DPEClient, t *testing.T) {
+func TestCertifyKeySimulation(d TestDPEInstance, c DPEClient, t *testing.T) {
```
Comment (hpya93) [inline verification/certifyKey.go:135]: Consistent naming

File: verification/certifyKey.go [hunk 5]
```diff
@@ -120,18 +123,23 @@ const (
 
 type TcgMultiTcbInfo = []DiceTcbInfo
 
+type CertifyKeyParams struct {
+	Label []byte
+	Flags CertifyKeyFlags
+}
+
```
Comment (hpya93) [inline verification/certifyKey.go:130]: Moved Params struct ouside the function to reduce size

File: verification/certifyKey.go [hunk 6]
```diff
@@ -249,42 +262,55 @@ func checkCertifyKeyMultiTcbInfoExtensionStructure(t *testing.T, c *x509.Certifi
...
-		hasher.Write(defaultTci)
 	}
 
-	// The DiceTcbInfo blocks are listed with current node at index0 followed by parent TCI nodes.
-	for i := len(multiTcbInfo) - 1; i >= 0; i-- {
-		hasher.Write(multiTcbInfo[i].Fwids[0].Digest)
+	// For Extended TCI, CUMULATIVE = H(TCI_CUMULATIVE || INPUT_TCI_VALUE)
+	startIndex := 0
+	if isTciExtended {
+		startIndex += 1 // To skip usual hash compuatation for extended TCI
```
Comment (hpya93) [inline verification/certifyKey.go:289]: Generally, TCI_CUM = H(DefaultTCI || TCI_CURRENT)
While extending derived context, TCI_CUM = H(H(DefaultTCI || TCI_CURRENT) || ExtendedTCI)

**Comment by jhand2** [comment]
Sorry about that, I also noticed this after the fact and added it back in https://github.com/chipsalliance/caliptra-dpe/pull/265.  265 also does some cleanup and uses some fixes I added in https://github.com/chipsalliance/caliptra-dpe/pull/262.

Running the tests in the Caliptra emulator revealed some edge cases that the existing ExtendTCI tests didn't cover, because Caliptra is auto-initialized with more TcbInfos.

**Comment by hpya93** [comment]
Its fine. In that case can I close this?

**Comment by hpya93** [comment]
> Its fine. In that case can I close this?

Closing the issue as it is already taken care in PR 262, 265
